### PR TITLE
bug fixing in animations and adding Geometric Objects

### PR DESCRIPTION
**Pull request by two-ticks**
## TODO
- [x] refactor play function
- [x] fix animation i.e. staggering, gradual fill etc.
- [ ] test cases for play and scene 
- [x] controls and timeline 
- [x] Geometric objects

### CODE REVIEW

**Comment by two-ticks** [comment]
**Feature** : `style(property, value)` method 
**Problem**: ~~works only when add or play has been called~~ (solved)
```js
  let title = createText("Pythagoras' Theorem")
  title.style('font-family', 'Courier New')
```

image

https://editor.p5js.org/radium.scientist/sketches/-XYDcX-B2
```js
  let title = createText("Pythagoras' Theorem")
  title.position(20, 50)
  title.fill(GOLD20)
  title.size(30)
  //title.style('font-family', 'Courier New') //will not work
  title.play('growFromCenter', 1, 2)
  title.style('font-family', 'Courier New') //works only after play or add has been called 
```

**Comment by two-ticks** [comment]
- [x]  fix overflowing

**Comment by two-ticks** [comment]
`clock()` returns current time of animation in milliseconds 

```
function draw() {
  t = clock(); //sets t = time of animation timeline
}
```
Description of example : position of rectangle is changed according to animation timeline
vA89EWIwIn

**Comment by two-ticks** [comment]
PR for improving GObject: #22

**Comment by two-ticks** [comment]
- [ ] update API for GObjects
- [ ] test manually animations
- [ ] add more customization method

**Comment by two-ticks** [comment]
## `create2DGraph(equation, x, y, width, height)`

**Parameters**

| Parameter | Type        | Description            |
| :-------- | :---------- | :--------------------- |
| equation  | `function`  | function to be plotted |
| [x]       | `number`    | x-coordinate of graph  |
| [y]       | `number`    | y-coordinate of graph  |
| [width]   | `number` px | width of SVG           |
| [height]  | `number` px | height of SVG          |

[ ] : optional arguments

```js
curve = create2DGraph((t) => 400 + 1500 * Math.cos(2 * t));
);
```

| Method                     | Description                           |
| :------------------------- | :------------------------------------ |
| object.plot()              | plots the graph                       |
| object.position(x, y)      | sets position of graph object         |
| object.size(width, height) | sets width and height of graph object |
| object.stroke(strokeColor) | sets stroke color of graph object     |
| object.play()              | play animation of creation            |
| object1.transform(object2) | transforms object1 into object2       |

**Example**

[🔗example sketch](https://editor.p5js.org/radium.scientist/sketches/Nl55ATBHk)

<details>
  <summary>Code</summary>

```js
function setup() {
  createCanvas(400, 400);
  background(220);
  scene = new Scene();
  reel();
}

function draw() {
  background(220);
}

function reel() {
  grp = create2DGraph(
    (t) =>
      800 * Math.cos(2 * t) + 1000 * Math.sin(4 * t) + 800 * Math.sin(6 * t)
  );

  grp.plot();
  grp.position(10, 200);
  grp.play();

  grp1 = create2DGraph((t) => 800 * Math.cos(2 * t));
  grp1.plot();
  grp1.position(10, 0);
  grp1.play();

  grp2 = create2DGraph((t) => 1000 * Math.sin(4 * t));
  grp2.plot();
  grp2.position(10, 50);
  grp2.play();

  grp3 = create2DGraph((t) => 800 * Math.sin(6 * t));
  grp3.plot();
  grp3.position(10, 100);
  grp3.play();
}
```

</details>

## `create2DPolarGraph(equation, theta-range, x, y, width, height)`

**Parameters**

| Parameter   | Type           | Description            |
| :---------- | :------------- | :--------------------- |
| equation    | `function`     | function to be plotted |
| theta-range | `number array` | range of theta         |
| [x]         | `number`       | x-coordinate of graph  |
| [y]         | `number`       | y-coordinate of graph  |
| [width]     | `number` px    | width of SVG           |
| [height]    | `number` px    | height of SVG          |

[ ] : optional arguments

```js
curve = create2DPolarGraph(
  (t) =>
    15 *
    (Math.pow(Math.E, Math.cos(t)) -
      2 * Math.cos(4 * t) -
      Math.pow(Math.sin(t / 12), 5)),
  [0, 12 * Math.PI]
);
```

| Method                     | Description                           |
| :------------------------- | :------------------------------------ |
| object.plot()              | plots the graph                       |
| object.position(x, y)      | sets position of graph object         |
| object.size(width, height) | sets width and height of graph object |
| object.stroke(strokeColor) | sets stroke color of graph object     |
| object.play()              | play animation of creation            |
| object1.transform(object2) | transforms object1 into object2       |

**Example**

[🔗example sketch](https://editor.p5js.org/radium.scientist/sketches/am40f47oTj)

<details>
  <summary>Code</summary>

```js
function setup() {
  createCanvas(400, 400);
  background(220);
  scene = new Scene();
  reel();
}

function draw() {
  background(220);
}

function reel() {
  grp1 = create2DPolarGraph(
    (t) =>
      15 *
      (Math.pow(Math.E, Math.cos(t)) -
        2 * Math.cos(4 * t) -
        Math.pow(Math.sin(t / 12), 5)),
    [0, 12 * Math.PI]
  );
  grp1.plot();
  grp1.size(400, 400);
  grp1.position(50, 50);
  grp1.play();
}
```

</details>

## `create2DParametricGraph(x, y, [range])`

**Parameters**

| Parameter    | Type           | Description        |
| :----------- | :------------- | :----------------- |
| x expression | `function`     | x expression       |
| y expression | `function`     | y expression       |
| [range]      | `number array` | range of parameter |

[ ] : optional arguments

```js
curve = create2DParametricGraph(
  (t) => 40 * Math.sin(4 * t + Math.PI / 4),
  (t) => 40 * Math.cos(5 * t)
);
```

| Method                     | Description                           |
| :------------------------- | :------------------------------------ |
| object.plot()              | plots the graph                       |
| object.position(x, y)      | sets position of graph object         |
| object.size(width, height) | sets width and height of graph object |
| object.stroke(strokeColor) | sets stroke color of graph object     |
| object.play()              | play animation of creation            |
| object1.transform(object2) | transforms object1 into object2       |

**Example**

[🔗example sketch](https://editor.p5js.org/radium.scientist/sketches/_CHFblWtj)

<details>
  <summary>Code</summary>

```js
let button;
let A = 40;
let B = 40;
let a = 3;
let b = 2;
let d = Math.PI / 4;

function setup() {
createCanvas(400, 400);
background(0);
scene = new Scene();
createControls();
reel();
}

function reel() {
curve1 = create2DParametricGraph(
(t) => A _ Math.sin(a _ t + d),
(t) => B _ Math.cos(b _ t)
);

curve1.plot();
curve1.stroke('blue');
curve1.position(50, 50);
a = 5;
b = 4;
d = Math.PI / 8;

curve2 = create2DParametricGraph(
(t) => A _ Math.sin(a _ t + d),
(t) => B _ Math.cos(b _ t)
);
curve2.stroke('blue');
// curve2.plot();
curve2.size(400, 400);
curve2.position(50, 150);
curve1.transform(curve2, 0, 10);
}

```

</details>

**Comment by two-ticks** [comment]
Should we consider merging #20 to `develop` and before starting to work on #23 issue ?  

@nickmcintyre @JithinKS97

File: api_reference.md [hunk 1]
```diff
@@ -137,3 +160,802 @@ async function reel() {
...
+| equation  | `function`  | function to be plotted |
+| [x]       | `number`    | x-coordinate of graph  |
+| [y]       | `number`    | y-coordinate of graph  |
+| [width]   | `number` px | width of SVG           |
+| [height]  | `number` px | height of SVG          |
+
+[ ] : optional arguments
+
+```js
+curve = create2DGraph((t) => 400 + 1500 * Math.cos(2 * t));
```
Comment (nickmcintyre) [inline api_reference.md]: Let's default to using p5's [math functions](https://p5js.org/reference/#group-Math) where possible.

Comment (two-ticks) [reply api_reference.md]: this is great!

**Review by nickmcintyre — CHANGES_REQUESTED** [review]
@two-ticks truly outstanding work thus far! One spot in the API stood out to me:

```javascript
scene = new Scene();
```

Given the `scene` variable is never used directly, can we handle this task automatically by wrapping the statement in a function and [registering it with p5](https://github.com/processing/p5.js/blob/main/contributor_docs/creating_libraries.md#use-registermethod-to-register-functions-with-p5-that-should-be-called-at-various-times)?

@JithinKS97 I made a few tweaks to style and stopped short of removing commented-out code for clarity. Curious to hear your thoughts on style and testing.

**Comment by two-ticks** [comment]
- [x] registered `scene` with p5
- [x] added `overflow` function 

index.ts
```diff
- import { Scene } from './lib/Scene/scene';
+ import { Scene, overflow } from './lib/Scene/scene';
+ global.overflow = overflow;
+ global.p5.prototype.registerMethod('init', Scene);
```
sketch.js
```diff
function setup() {
  canvas = createCanvas(400, 400);
- scene = new Scene();
+ //overflow('hidden'); //if you want to stop DOM p5 teach elements from overflowing, cannot be called before canvas
  background(20);
  drawGraph();
}
```

**Comment by JithinKS97** [comment]
@nickmcintyre The changes looks good. The documentation / description that you added will definitely help.

**Comment by two-ticks** [comment]
Should we merge `beginGraph` in `GObject` ?  #25

**Review by nickmcintyre — APPROVED** [review]
Great work @two-ticks! The library is really becoming a joy to use.

**Pull request merged**